src/nvim/tui/input.c

```diff
@@ -134,12 +132,39 @@ static void tinput_wait_enqueue(void **argv)
   uv_mutex_unlock(&input->key_buffer_mutex);
 }
 
+static void tinput_paste_event(void **argv)
+{
+  static bool canceled = false;
+
+  String keys = { .data = argv[0], .size = (size_t)argv[1] };
+  intptr_t phase = (intptr_t)argv[2];
+
+  if (phase == -1 || phase == 1) {
+    canceled = false;
+  }
+
+  Error err = ERROR_INIT;
+  if (!canceled) {
+    if (!nvim_paste(keys, true, phase, &err)) {
+      // paste failed, ingore further segments of the same paste
+      canceled = true;
```

Comment:
"cancel" is different than "fail". NB `nvim_paste` was intended/designed to discard the chunks, so that TUI could avoid needing extra logic to "ignore further segments".

Reply:
I just tried to mirror the existing logic which cancels the paste on `nvim_paste` returning false. But this logic was not necessary?

Reply:
I guess it is necessary now because the `input->paste` state can't be updated on the TUI thread. So we need this extra state now.

https://github.com/neovim/neovim/commit/f5fd699c527945d795dab9f3529333df6d0561d0 adds some coverage for the cancel case.

Reply:
@justinmk It seems to work by now. But I separated https://github.com/neovim/neovim/commit/2c605d1f22a243bc34b680f69c7c8cfe01b80887 so it easily can be reverted if a test shows that it is necessary.